adal/src/main/java/com/microsoft/aad/adal/Oauth2.java

```diff
@@ -590,17 +602,29 @@ public String encodeProtocolState() throws UnsupportedEncodingException {
      * @param webResponse the web response from which authentication result will be constructed
      * @return AuthenticationResult
      */
-    private AuthenticationResult processTokenResponse(HttpWebResponse webResponse) throws AuthenticationException {
+    private AuthenticationResult processTokenResponse(HttpWebResponse webResponse, final HttpEvent httpEvent)
+            throws AuthenticationException {
         AuthenticationResult result;
         String correlationIdInHeader = null;
-        if (webResponse.getResponseHeaders() != null
-                && webResponse.getResponseHeaders().containsKey(
-                        AuthenticationConstants.AAD.CLIENT_REQUEST_ID)) {
-            // headers are returning as a list
-            List<String> listOfHeaders = webResponse.getResponseHeaders().get(
-                    AuthenticationConstants.AAD.CLIENT_REQUEST_ID);
-            if (listOfHeaders != null && listOfHeaders.size() > 0) {
-                correlationIdInHeader = listOfHeaders.get(0);
+        if (webResponse.getResponseHeaders() != null) {
+            if (webResponse.getResponseHeaders().containsKey(
+                    AuthenticationConstants.AAD.CLIENT_REQUEST_ID)) {
+                // headers are returning as a list
+                List<String> listOfHeaders = webResponse.getResponseHeaders().get(
+                        AuthenticationConstants.AAD.CLIENT_REQUEST_ID);
+                if (listOfHeaders != null && listOfHeaders.size() > 0) {
+                    correlationIdInHeader = listOfHeaders.get(0);
+                }
+            }
+
+            if (webResponse.getResponseHeaders().containsKey(AuthenticationConstants.AAD.REQUEST_ID_HEADER)) {
+                // headers are returning as a list
+                List<String> listOfHeaders = webResponse.getResponseHeaders().get(
+                        AuthenticationConstants.AAD.CLIENT_REQUEST_ID);
```

Comment:
Oops, copy paste. Good catch! i will test this before sending it out

---
In reply to: [90366111](https://github.com/AzureAD/azure-activedirectory-library-for-android/pull/736#discussion_r90366111) [](ancestors = 90366111)